controlplane-rs/src/gateway_controller.rs

```diff
@@ -0,0 +1,178 @@
+        return Ok(Action::requeue(Duration::from_secs(3600)));
+    }
+    debug!(
+        "found a supported GatewayClass: {:?}",
+        gateway_class.name_any()
+    );
+
+    set_listener_status(&mut gw);
+    let accepted_cond = get_accepted_condition(&mut gw);
+    set_condition(&mut gw, accepted_cond.clone());
+    if accepted_cond.status == "False".to_string() {
+        let programmed_cond = metav1::Condition {
+            last_transition_time: accepted_cond.last_transition_time.clone(),
+            observed_generation: accepted_cond.observed_generation,
+            type_: "Programmed".to_string(),
+            status: "False".to_string(),
+            message: accepted_cond.message.clone(),
+            reason: "Invalid".to_string(),
+        };
+        set_condition(&mut gw, programmed_cond);
+    }
+
+    if accepted_cond.status == String::from("False") {
+        patch_status(&gateway_api, name, gw.status.as_ref().unwrap()).await?;
+        return Err(Error::InvalidConfigError(accepted_cond.message));
+    }
+
+    let service_api: Api<Service> = Api::namespaced(client, &ns);
+    let services = service_api
+        .list(&ListParams::default().labels(&format!("{}={}", GATEWAY_SERVICE_LAEBL, name)))
+        .await
+        .map_err(Error::KubeError)?;
+    if services.items.len() > 1 {
+        return Err(Error::LoadBalancerError(
+            "found more than 1 Service for this Gateway; multiple services are not supported"
+                .to_string(),
+        ));
+    }
+
+    let mut service: Service;
+    if let Some(val) = services.items.get(0) {
+        service = val.clone();
+        let updated = update_service_for_gateway(gateway.as_ref(), &mut service)?;
+        if updated {
+            info!("drift detected; updating loadbalancer service");
+            let patch_parmas = PatchParams::default();
+            service_api
+                .patch(
+                    val.name_any().as_str(),
+                    &patch_parmas,
+                    &Patch::Strategic(&service),
+                )
+                .await
+                .map_err(Error::KubeError)?;
+        }
+    } else {
+        info!("creating loadbalancer service");
+        service = create_svc_for_gateway(ctx.clone(), gateway.as_ref()).await?;
+    }
+
+    let now = k8s_openapi::apimachinery::pkg::apis::meta::v1::Time(Utc::now());
+    if get_ingress_ip_len(&service) == 0 || service.spec.as_ref().unwrap().cluster_ip.is_none() {
+        let msg = "LoadBalancer does not have a ingress IP address".to_string();
+        set_condition(
+            &mut gw,
+            metav1::Condition {
+                last_transition_time: now,
+                observed_generation: gateway.meta().generation,
+                message: msg.clone(),
+                reason: "AddressNotAssigned".to_string(),
+                status: "False".to_string(),
+                type_: "Programmed".to_string(),
```

Comment:
Similar to above with getting our conditions from the `gateway-api` crate if we can.